File: src/schemas/metadata/MetadataParameter.ts
import ts from "typescript";

import { ClassProperties } from "../../typings/ClassProperties";

import { IJsDocTagInfo } from "./IJsDocTagInfo";
import { IMetadataDictionary } from "./IMetadataDictionary";
import { IMetadataParameter } from "./IMetadataParameter";
import { Metadata } from "./Metadata";

export class MetadataParameter {
  public name: string;
  public type: Metadata;
  public description: string | null;
  public jsDocTags: IJsDocTagInfo[];
  public tsType?: ts.Type;

  private constructor(props: ClassProperties<MetadataParameter>) {
    this.name = props.name;
    this.type = props.type;
    this.description = props.description;
    this.jsDocTags = props.jsDocTags;
    this.tsType = props.tsType;
  }

  /** @internal */
  public static create(
    props: ClassProperties<MetadataParameter>,
  ): MetadataParameter {
    return new MetadataParameter(props);
  }

  public static from(
    json: IMetadataParameter,
    dict: IMetadataDictionary,
  ): MetadataParameter {
    return MetadataParameter.create({
      name: json.name,
      type: Metadata.from(json.type, dict),
      description: json.description,
      jsDocTags: json.jsDocTags,
    });
  }

  public toJSON(): IMetadataParameter {
    return {
      name: this.name,
      type: this.type.toJSON(),
      description: this.description,
      jsDocTags: this.jsDocTags,
    };
  }
}

File: packages/typia/native/core/schemas/metadata/MetadataParameter.go
package metadata

import shimchecker "github.com/microsoft/typescript-go/shim/checker"

type IMetadataSchema_IParameter struct {
	Name        string
	Type        *IMetadataSchema
	Description *string
	JsDocTags   []IJsDocTagInfo
}

type MetadataParameter struct {
	Name        string
	Type        *MetadataSchema
	Description *string
	JsDocTags   []IJsDocTagInfo
	TsType      *shimchecker.Type
}

func MetadataParameter_create(props MetadataParameter) *MetadataParameter {
	return &MetadataParameter{
		Name:        props.Name,
		Type:        props.Type,
		Description: props.Description,
		JsDocTags:   append([]IJsDocTagInfo{}, props.JsDocTags...),
		TsType:      props.TsType,
	}
}

func MetadataParameter_from(json IMetadataSchema_IParameter, dict IMetadataDictionary) *MetadataParameter {
	return MetadataParameter_create(MetadataParameter{
		Name:        json.Name,
		Type:        MetadataSchema_from(json.Type, dict),
		Description: json.Description,
		JsDocTags:   json.JsDocTags,
	})
}

func (obj *MetadataParameter) ToJSON() IMetadataSchema_IParameter {
	return IMetadataSchema_IParameter{
		Name:        obj.Name,
		Type:        obj.Type.ToJSON(),
		Description: obj.Description,
		JsDocTags:   append([]IJsDocTagInfo{}, obj.JsDocTags...),
	}
}
